clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -2507,6 +2507,7 @@ static void CollectArgsForIntegratedAssembler(Compilation &C,
   bool Crel = false, ExperimentalCrel = false;
   bool UseRelaxRelocations = C.getDefaultToolChain().useRelaxRelocations();
   bool UseNoExecStack = false;
+  bool Msa = false;
```

Comment:
`std::optional<bool> Msa`. `-mno-msa` should lead to an error even on non-mips targets.

Reply:
I added bool NoMsa and lead to "unsupported" error when it is on non-mips, were I understanding correctly?

Reply:
`bool Msa, NoMsa` encodes 4 states where `Msa&&NoMsa` is meaningless but not checked.

`std::optional<bool> Msa` encodes the desired 3 states:

* not set
* set and true
* set and false

Reply:
OK, I updated.

Reply:
Please do use `std::optional<bool> Msa` and please don't add `err_drv_msa_and_nomsa`. It's common for `-Wa,--foo` to override a previous `-Wa,--no-foo`. There shouldn't be an error.

Reply:
> `bool Msa, NoMsa` encodes 4 states where `Msa&&NoMsa` is meaningless but not checked.

What does this mean？
I thought you were referring to bool Msa && bool NoMsa.

Reply:
```
std::optional<bool> Msa

if (-mmsa) Msa = true;
if (-mno-msa) Msa = false;


if (Msa) {
  if (not mips) ...
  else if (*Msa) CmdArgs.push_back("-mmsa");
}
```

Reply:
> `std::optional<bool> Msa`. `-mno-msa` should lead to an error even on non-mips targets.
What error need I add in this https://github.com/llvm/llvm-project/commit/444cd83d070600a26938514070d79dc900946868, and what is *Msa?

Reply:
>  `-mno-msa` should lead to an error even on non-mips targets.

Assemblers accept this option, so I don't think we should throw an error when `-mno-msa` is used with MIPS targets.

Reply:
> `bool Msa, NoMsa` encodes 4 states where `Msa&&NoMsa` is meaningless but not checked.
> 
> `std::optional<bool> Msa` encodes the desired 3 states:
> 
> * not set
> * set and true
> * set and false

Sorry, I didn't realize you were asking me to use an option type bool MSA on the first version of the modification.

Reply:
BTW using `std::Optional<bool>` makes more sense since it does distinguish between `None` (totally not set) and Boolean `false` (disabled). Your initial thought is that `false` stands for 1) MSA flag is not set and 2) MSA is disabled, but if we target a non-MIPS machine, then we must know if the `-mno-msa` is actually set - that's why using plain `bool` in this context is not a good idea, `false` can not mean two states.